Add an entry in the Formants menu to output the listing of formant
bandwidths.

 -- Rafael Laboissiere <rafael@debian.org>  Sun, 22 Feb 2009 12:45:41 +0100

--- a/fon/FunctionEditor_SoundAnalysis.c
+++ b/fon/FunctionEditor_SoundAnalysis.c
@@ -1907,6 +1907,7 @@
 	EditorMenu_addCommand (menu, L"-- query formants --", 0, NULL);
 	EditorMenu_addCommand (menu, L"Query:", motif_INSENSITIVE, menu_cb_getFrequency /* dummy */);
 	EditorMenu_addCommand (menu, L"Formant listing", 0, menu_cb_formantListing);
+	EditorMenu_addCommand (menu, L"Bandwidth listing", 0, menu_cb_bandwitdhtListing);
 	EditorMenu_addCommand (menu, L"Get first formant", motif_F1, menu_cb_getFirstFormant);
 	EditorMenu_addCommand (menu, L"Get first bandwidth", 0, menu_cb_getFirstBandwidth);
 	EditorMenu_addCommand (menu, L"Get second formant", motif_F2, menu_cb_getSecondFormant);
--- a/fon/TimeSoundAnalysisEditor.c
+++ b/fon/TimeSoundAnalysisEditor.c
@@ -1330,6 +1330,42 @@
 	return 1;
 }
 
+static int menu_cb_bandwitdhtListing (EDITOR_ARGS) {
+	EDITOR_IAM (TimeSoundAnalysisEditor);
+	double tmin, tmax;
+	int part = makeQueriable (me, TRUE, & tmin, & tmax); iferror return 0;
+	if (! my formant.show)
+		return Melder_error1 (L"No formant contour is visible.\nFirst choose \"Show formants\" from the Formant menu.");
+	if (! my formant.data) {
+		TimeSoundAnalysisEditor_computeFormants (me);
+		if (! my formant.data) return Melder_error1 (theMessage_Cannot_compute_formant);
+	}
+	MelderInfo_open ();
+	MelderInfo_writeLine1 (L"Time_s   F1_Hz   F2_Hz   F3_Hz   F4_Hz");
+	if (part == FunctionEditor_PART_CURSOR) {
+		double f1 = Formant_getBandwidthAtTime (my formant.data, 1, tmin, 0);
+		double f2 = Formant_getBandwidthAtTime (my formant.data, 2, tmin, 0);
+		double f3 = Formant_getBandwidthAtTime (my formant.data, 3, tmin, 0);
+		double f4 = Formant_getBandwidthAtTime (my formant.data, 4, tmin, 0);
+		MelderInfo_write5 (Melder_fixed (tmin, 6), L"   ", Melder_fixed (f1, 6), L"   ", Melder_fixed (f2, 6));
+		MelderInfo_writeLine4 (L"   ", Melder_fixed (f3, 6), L"   ", Melder_fixed (f4, 6));
+	} else {
+		long i, i1, i2;
+		Sampled_getWindowSamples (my formant.data, tmin, tmax, & i1, & i2);
+		for (i = i1; i <= i2; i ++) {
+			double t = Sampled_indexToX (my formant.data, i);
+			double f1 = Formant_getBandwidthAtTime (my formant.data, 1, t, 0);
+			double f2 = Formant_getBandwidthAtTime (my formant.data, 2, t, 0);
+			double f3 = Formant_getBandwidthAtTime (my formant.data, 3, t, 0);
+			double f4 = Formant_getBandwidthAtTime (my formant.data, 4, t, 0);
+			MelderInfo_write5 (Melder_fixed (t, 6), L"   ", Melder_fixed (f1, 6), L"   ", Melder_fixed (f2, 6));
+			MelderInfo_writeLine4 (L"   ", Melder_fixed (f3, 6), L"   ", Melder_fixed (f4, 6));
+		}
+	}
+	MelderInfo_close ();
+	return 1;
+}
+
 static int do_getFormant (TimeSoundAnalysisEditor me, int iformant) {
 	double tmin, tmax;
 	int part = makeQueriable (me, TRUE, & tmin, & tmax); iferror return 0;
@@ -1666,6 +1702,7 @@
 	EditorMenu_addCommand (menu, L"-- formant query --", 0, NULL);
 	EditorMenu_addCommand (menu, L"Query:", GuiMenu_INSENSITIVE, menu_cb_getFrequency /* dummy */);
 	EditorMenu_addCommand (menu, L"Formant listing", 0, menu_cb_formantListing);
+	EditorMenu_addCommand (menu, L"Bandwith listing", 0, menu_cb_bandwitdhtListing);
 	EditorMenu_addCommand (menu, L"Get first formant", GuiMenu_F1, menu_cb_getFirstFormant);
 	EditorMenu_addCommand (menu, L"Get first bandwidth", 0, menu_cb_getFirstBandwidth);
 	EditorMenu_addCommand (menu, L"Get second formant", GuiMenu_F2, menu_cb_getSecondFormant);
